R/block_cv.R

```diff
@@ -0,0 +1,172 @@
+#' grid cells, or "blocks", and then assigns all data into folds based on the
+#' blocks they fall into.
+#'
+#' @details
+#' The grid blocks can be controlled by passing arguments to
+#' [sf::st_make_grid()] via `...`. Some particularly useful arguments include:
+#'
+#' * `cellsize` Target cellsize, expressed as the "diameter" (shortest
+#' straight-line distance between opposing sides; two times the apothem)
+#' of each block, in map units.
+#' * `n` The number of grid blocks in the x and y direction (columns, rows).
+#' * `square` A logical value indicating whether to create square (`TRUE`) or
+#' hexagonal (`FALSE`) cells.
+#'
+#' If both `cellsize` and `n` are provided, then the number of blocks requested
+#' by `n` of sizes specified by `cellsize` will be returned, likely not
+#' lining up with the bounding box of `data`. If only `cellsize`
+#' is provided, this function will return as many blocks of size
+#' `cellsize` as fit inside the bounding box of `data`. If only `n` is provided,
+#' then `cellsize` will be automatically adjusted to create the requested
+#' number of cells.
+#'
+#' @param data An object of class `sf` or `sfc`.
+#' @param method The method used to sample blocks for cross validation folds.
+#' Currently, only `"random"` is supported.
+#' @inheritParams rsample::vfold_cv
+#' @param ... Arguments passed to [sf::st_make_grid()].
+#'
+#' @return A tibble with classes `spatial_block_cv`, `rset`, `tbl_df`, `tbl`,
+#'   and `data.frame`. The results include a column for the data split objects
+#'   and an identification variable `id`.
+#'
+#' @examples
+#' data(Smithsonian, package = "modeldata")
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
+#'                                coords = c("longitude", "latitude"),
+#'                                # Set CRS to WGS84
+#'                                crs = 4326)
+#'
+#' spatial_block_cv(smithsonian_sf, v = 3)
+#'
+#' @references
+#'
+#' D. R. Roberts, V. Bahn, S. Ciuti, M. S. Boyce, J. Elith, G. Guillera-Arroita,
+#' S. Hauenstein, J. J. Lahoz-Monfort, B. Schröder, W. Thuiller, D. I. Warton,
+#' B. A. Wintle, F. Hartig, and C. F. Dormann. "Cross-validation strategies for
+#' data with temporal, spatial, hierarchical, or phylogenetic structure," 2016,
+#' Ecography 40(8), pp. 913-929, doi: 10.1111/ecog.02881.
+#'
+#' @export
+spatial_block_cv <- function(data, method = "random", v = 10, ...) {
+  method <- rlang::arg_match(method)
+
+  if (!"sf" %in% class(data)) {
+    rlang::abort(
+      c(
+        "`spatial_block_cv()` currently only supports `sf` objects.",
+        i = "Try converting `data` to an `sf` object via `sf::st_as_sf()`."
+      )
+    )
+  }
+
+  if (sf::st_crs(data) == sf::NA_crs_) {
+    rlang::abort(
+      c(
+        "`spatial_block_cv()` requires your data to have an appropriate coordinate reference system (CRS).",
+        i = "Try setting a CRS using `sf::st_set_crs()`."
+      )
+    )
+  }
```

Comment:
If you find yourself copy/pasting this for other resampling methods for sf, let's make this a little function.